libcudf Developer Guide [skip-ci]

harrism opened the pull request:
Adds a new developer guide for libcudf. This is based on the
existing libcudf++ transition guide.

Fixes #5273 

TODO

- [x] Description of `dictionary_column_wrapper` and `fixed_point_column_wrapper`
- [x] Benchmarking Section (put in a new file, Benchmarking.md)?
- [x] Better discussion of nested types
- [x] Introductory section on data types
- [x] Consider splitting into multiple documents: DEVELOPER_GUIDE.md, TESTING.md, BENCHMARKING.md?
- [x] Placeholder for cuIO?
- [x] Add section on code and documentation style and formatting

jrhemstad commented on cpp/docs/DEVELOPER_GUIDE.md:
````diff
@@ -0,0 +1,981 @@
+and produce `unique_ptr`s to owning objects as output. For example, 
+
+```c++
+std::unique_ptr<table> sort(table_view input);
````
```suggestion
std::unique_ptr<table> sort(table_view const& input);
```

jrhemstad commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,981 @@
+with the `*_view` suffix is non-owning. For more detail see the 
+[`libcudf++` presentation.](https://docs.google.com/presentation/d/1zKzAtc1AWFKfMhiUlV5yRZxSiPLwsObxMlWRWz_f5hA/edit?usp=sharing)
+
+`libcudf` functions typically take views as input (`column_view`, `table_view`, or `scalar_view`)
```
```suggestion
`libcudf` functions typically take views as input (`column_view` or `table_view`)
```

harrism replied:
We don't have `scalar_view`?

jrhemstad commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,981 @@
+
+## `rmm::device_memory_resource`<a name="memory_resource"></a>
+
+`libcudf` Allocates all memory via RMM memory resources (MR). See the 
```
```suggestion
`libcudf` Allocates all device memory via RMM memory resources (MR). See the 
```

jrhemstad commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,981 @@
+    // defaulted stream parameter
+    void external_function(..., rmm::cuda_stream_view stream){
+        // implementation uses stream w/ async APIs
+        RMM_ALLOC(...,stream);
```
yikes
```suggestion
        rmm::device_buffer buff(...,stream);
```

harrism replied:
Note that I have only edited the very beginning of this document so far. This is a WIP PR not really ready for review.

It *is* ready for you to push directly to the PR though if you would like.

jrhemstad commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,981 @@
+        kernel<<<..., stream>>>(...);
+        thrust::algorithm(rmm::exec_policy(stream), ...);
+        stream.synchronize();
+        RMM_FREE(...,stream);
```
```suggestion
```

jrhemstad commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,981 @@
+        CUDA_TRY(cudaMemcpyAsync(...,stream.value()));
+        kernel<<<..., stream>>>(...);
+        thrust::algorithm(rmm::exec_policy(stream), ...);
+        stream.synchronize();
```
This sync may lead people to believe that all detail functions should sync their stream before returning.

jrhemstad commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,981 @@
+
+### Memory Management
+
+Explicit memory management through calls to `RMM_ALLOC/RMM_FREE` should be avoided whenever possible in favor of a construct with automated lifetime management, i.e., a RAII object. 
```
```suggestion
```

jrhemstad commented on cpp/docs/DEVELOPER_GUIDE.md:
````diff
@@ -0,0 +1,981 @@
+int host_value = int_scalar.value();
+```
+
+#### `rmm::device_vector<T>`
+
+Allocates a specified number of elements of the specified type.
+If no initialization value is provided, all elements are default initialized (this incurs a kernel launch).
+
+(*) Note: `rmm::device_vector<T>` is not yet updated to use `device_memory_resource`s, but support is forthcoming.
````
Should add a section for `device_uvector`.

jrhemstad commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,981 @@
+For most functions, the top-level `cudf` namespace is sufficient. However, for logically grouping a broad set of functions, further namespaces may be used. For example, there are numerous functions that are specific to columns of Strings. These functions are put in the `cudf::strings::` namespace. Similarly, functionality used exclusively for unit testing is placed in the `cudf::test::` namespace ...
+
+
+### `experimental`
+During the transition period, symbols in `libcudf++` that conflict with old symbol names should be placed in the `cudf::experimental` namespace to prevent collision with the old symbols, e.g., `cudf::experimental::table` and `cudf::experimental::type_dispatcher`. Once the transition is complete, the `experimental` namespace will be removed.
```
```suggestion
```

jrhemstad commented on cpp/docs/DEVELOPER_GUIDE.md:
````diff
@@ -0,0 +1,981 @@
+// This will always invoke `operator()<int32_t>`
+cudf::type_dispatcher<always_int>(data_type, f);
+```
+
````
```suggestion
## Specializing Type Dispatched Code Paths
```

jrhemstad commented on cpp/docs/DEVELOPER_GUIDE.md:
````diff
@@ -0,0 +1,981 @@
+auto validity = make_counting_transform_iterator(0, [](auto i){return i%2;});
+strings_column_wrapper s({"", "this", "is", "a", "column", "of", "strings"}, validity);
+```
+
````
Should add a section on list_column_wrapper. @nvdbaranec

harrism replied:
Done.

jrhemstad commented on cpp/docs/DEVELOPER_GUIDE.md:
````diff
@@ -0,0 +1,981 @@
+strings_column_wrapper s({"", "this", "is", "a", "column", "of", "strings"}, validity);
+```
+
+
````
Should add a section on `fixed_point_column_wrapper`. @codereport

harrism replied:
Done.

jrhemstad commented on cpp/docs/DEVELOPER_GUIDE.md:
````diff
@@ -0,0 +1,981 @@
+All that needs to be changed in Cython are the paths to legacy headers.
+For example, after moving `cudf/copying.hpp` to `cudf/legacy/copying.hpp`, the following line in
+`python/cudf/cudf/_lib/includes.copying.pxd` will be changed from:
+
+```python
+cdef extern from "cudf/copying.hpp" namespace "cudf" nogil:
+```
+
+to:
+
+```python
+cdef extern from "cudf/legacy/copying.hpp" namespace "cudf" nogil:
+```
+
+## Common Replacements<a name="replacements"></a>
+
+### Data Structures
+|          Old         |                New                |                      Notes                      |
+|:--------------------:|:---------------------------------:|:-----------------------------------------------:|
+|  `gdf_column const&` |        `cudf::column_view`        |                                                 |
+|    `gdf_column *`    |    `cudf::mutable_column_view`    | Only use when in/out is actually needed (rare). |
+|  `gdf_column const*` |        `cudf::column_view`        |                                                 |
+| `cudf::table const&` |         `cudf::table_view`        |                                                 |
+|    `cudf::table *`   |     `cudf::mutable_table_view`    |                                                 |
+| `cudf::device_table` |     `cudf::table_device_view`     |                    Immutable                    |
+| `cudf::device_table` | `cudf::mutable_table_device_view` |                     Mutable                     |
+| `gdf_scalar`         | `cudf::scalar`                    |                     Not yet implemented         |
+
+### Functions
+
+|              Old             |                       New                       |                                     Notes                                    |
+|:----------------------------:|:-----------------------------------------------:|:----------------------------------------------------------------------------:|
+|       `gdf_is_valid()`       |        `*column_device_view::is_valid()`        |         Verifies existence of bitmask before  attempting to read it.         |
+|    `bit_mask::is_valid()`    |    `*column_device_view::is_valid_nocheck()`    | Does not verify the existence of the bitmask  before attempting to read it.  |
+| `gdf_vald_allocation_size()` |        `bitmask_allocation_size_bytes()`        |                                                                              |
+|        `is_nullable()`       |               `*view::nullable()`               |                                                                              |
+|         `has_nulls()`        |               `*view::has_nulls()`              |                                                                              |
+|        `cudf::copy()`        | `column::column(const&)` `table::table(const&)` |                               Copy constructors                              |
+|    `cudf::allocate_like()`   |      `cudf::allocate_like()`      |                      This applies to both columns/tables                     |
+|     `cudf::empty_like()`     |        `cudf::empty_like()`       |                      This applies to both columns/tables                     |
````
All this can be deleted.

jrhemstad commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,981 @@
+
+## Strings Support<a name="string_support"></a>
+
+One of the more significant changes in `libcudf++` is the addition of native support for columns of strings. 
+Previously, string support in libcudf was accomplished via the `NVCategory` class.
+See the [NVCategory](#nvcategory_changes) section for more detail. 
+This class will no longer exist in `libcudf++`, but a replacement "dictionary" column type is planned for the future to replace the functionality of NVCategory.
+
+All libcudf functions will now need to be able to operate on a column of string elements directly. 
+This poses a set of unique challenges, as unlike all previous libcudf types, the elements in a string column are variable width. 
```
```suggestion

```

jrhemstad commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,981 @@
+
+All libcudf functions will now need to be able to operate on a column of string elements directly. 
+This poses a set of unique challenges, as unlike all previous libcudf types, the elements in a string column are variable width. 
+In order to represent variable width strings, libcudf uses a *compound* column (i.e., column with children).
```
There should probably be a previous section generically describing column children. That can then lead into this section on strings. Then add a section on lists and structs.

jrhemstad commented on cpp/docs/DEVELOPER_GUIDE.md:
````diff
@@ -0,0 +1,981 @@
+The `string_view.cuh` also includes some utility methods for reading and writing (`to_char_utf8/from_char_utf8`) individual UTF-8 characters to/from byte arrays.
+
+
+### NVCategory<a name="nvcategory_changes"></a>
+
+Previously, libcudf's support for strings columns relied on the `NVCategory` class. 
+`NVCategory` was an object that emulated a *categorical* or *dictionary* column. A dictionary column is a compound column composed of a *dictionary* and *indices* into that dictionary. 
+The dictionary contains the unique set of values within the column.
+The indices contain the offset into the dictionary for each element in the column, e.g., element `i` is given by `dictionary[indices[i]]`.
+See the example below.
+
+Furthermore, the dictionary is sorted.
+This means the indices are order preserving, and relational comparisons between values in the indices column will provide the same ordering as if the strings were compared directly.
+For example, in the snippet below, `"green"` is ordered lexicographically before `"red"`, and `"green"`'s index is `0` where `"red"`'s is `1`.
+Since `0` is ordered before `1`, comparing `0` and `1` will provide the same ordering as comparing `"green"` and `"red"` directly.
+
+```c++
+strings:    {"red", "yellow", "red", "green", "yellow"}
+dictionary: {"green", "red", "yellow"}
+indices:    {1, 2, 1, 0, 2}
+```
+`libcudf` leveraged the order preserving nature of the indices to allow comparing string elements without ever having to compare strings directly. 
+In effect, this allowed us to treat columns of strings as fixed-width, numeric columns which greatly simplified handling of string columns. 
+
+However, there was extra work required to manage the `NVCategory` columns--all of which should be removed when porting to libcudf++.
+Any use of a function from `cudf/cpp/include/cudf/utilities/legacy/nvcategory_util.hpp` should be removed from the new implementation.
````
```suggestion
```

vuule commented:
What do you think about having a section for benchmarks? It would be similar to the tests section (just smaller, AFAICT) - framework, guidelines, utilities.

codereport commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,823 @@
+to these additional files for further documentation of `libcudf` best practices.
+
+* [Documentation Guide](DOCUMENTATION.md) for guidelines on documenting libcudf code.
+* [Testing Guide(TESTING.md) for guidelines on writing unit tests.
```
```suggestion
* [Testing Guide](TESTING.md) for guidelines on writing unit tests.
```

codereport commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,823 @@
+# Overview
+
+libcudf is a C++ library that provides GPU-accelerated data-parallel algorithms for processing 
+column-oriented tabular data. Algorithms provided range from slicing and filtering to sorting, 
+and various types of aggregations, to database type operations such as grouping and joins. libcudf
+serves a number of clients via multiple language interfaces, including Python and Java. Users may
```
The structure of this sentence is a bit odd "From x & y TO z & A TO types of B & C"
```suggestion
column-oriented tabular data. Algorithms provided range from slicing, filtering, sorting 
and various types of aggregations, to database type operations such as grouping and joins. libcudf
serves a number of clients via multiple language interfaces, including Python and Java. Users may
```

codereport commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,823 @@
+Header files should use the `#pragma once` include guard. 
+
+The naming of external API headers should be consistent with the name of the folder that contains 
+the source files that implement the API. For example, the implementation of the APIs found in  
+`cudf/cpp/include/cudf/copying.hpp` are located in `cudf/src/copying`. Likewise, the unit tests for 
```
There is an unnecessary line break here
```suggestion
the source files that implement the API. For example, the implementation of the APIs found in `cudf/cpp/include/cudf/copying.hpp` are located in `cudf/src/copying`. Likewise, the unit tests for 
```

codereport commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,823 @@
+
+Only use `.cu` and `.cuh` if necessary. A good indicator is the inclusion of `__device__` and other
+symbols that are only recognized by `nvcc`. Also, Thrust APIs executed on a device execution policy,
+which should always be `rmm::exec_policy` in libcudf.
```
```suggestion
which should always use `rmm::exec_policy` in libcudf.
```

codereport commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,823 @@
+
+Resource ownership is an essential concept in `libcudf`. In short, an "owning" object owns a 
+resource (such as device memory). It acquires that resource during construction and releases the 
+resource in destruction (RAII). A "non-owning" object does not own resources. Any class in `libcudf` 
```
```suggestion
resource in destruction ([RAII](https://en.cppreference.com/w/cpp/language/raii)). A "non-owning" object does not own resources. Any class in `libcudf` 
```

harrism commented:
I've marked this ready for review. I've also added a section on nested type columns. It's very high level, except for the detail on strings that @davidwendt had already written. @davidwendt , @nvdbaranec and @mythrocks could you please review this section (it's near the end)?

harrism commented:
@vuule I added a placeholder for cuIO. I think that part of the developer guide should be filled out (and moved to a separate file) in a follow up PR, not in this PR.

davidwendt commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,973 @@
+`rmm::cuda_stream_default`. The implementation should be wholly contained in the `detail` API 
+definition and use only asynchronous versions of CUDA APIs with the stream parameter.
+
+In order to make the `detail` API callable from other libcudf functions, it should be exposed in a 
+header placed in the `cudf/cpp/include/detail/` directory.
```
Seems we should mention this in the Directory Structure section as well?

harrism replied:
It's mentioned in the section on `detail` namespace, but added a sentence about where these detail headers should live to the Directory Structure section.

davidwendt commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,973 @@
+libcudf code generally eschews raw pointers and direct memory allocation. Use RMM classes built to
+use `device_memory_resource`(*)s for device memory allocation with automated lifetime management.
+
+#### `rmm::device_buffer`
```
Maybe these classes should be defined in the rmm repo developer guide and then just pointed to from here?

harrism replied:
We don't have such a guide yet. But when we do...

davidwendt commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,973 @@
+Dictionaries provide an efficient way to represent low-cardinality data by storing a single copy 
+of each value. A dictionary comprises a column of sorted keys and a column containing an index into the 
+keys column for each row of the parent column. The keys column may have any libcudf data type, 
+such as a numerical type or strings. The indices represent the corresponding positions of each 
+element's value in the keys.
```
```suggestion
such as a numerical type or strings. The indices are represent the corresponding positions of each 
element's value in the keys. The indices child column can be any unsigned integer type (UINT8, UINT16, UINT32, or UINT64).
```

harrism replied:
I don't think the added `are` is correct.

harrism replied:
Added the sentence about index types.

harrism commented on cpp/docs/TESTING.md:
```diff
@@ -0,0 +1,407 @@
+
+#### `strings_column_wrapper`
+
+The `strings_column_wrapper` class should be used to create columns of strings. It provides 
```
I like it better the other way, so I changed the two cases where `a column of` was used to `columns of`.

nvdbaranec commented on cpp/docs/TESTING.md:
```diff
@@ -0,0 +1,408 @@
+
+Note "equivalent type". Most types are equivalent if and only they are equal. `fixed_point` types
+are the exception. They are equivalent if the representation type is equal, even if they have 
+different scales.
```
One additional point here:

- Nested type columns can be equivalent in the case where they are both of size 0, but one of them has children (also empty) and the other does not.  For cases where size is > 0, both equals and equivalent expect the same # of children.

nvdbaranec commented on cpp/docs/TESTING.md:
```diff
@@ -0,0 +1,408 @@
+# Unit Testing in libcudf
```
I think it would be worth having a best practices section for testing.  There are a lot of classic things that are worthy of testing which could be easily overlooked:

- Anything that involves manipulating bitmasks (especially hand-rolled kernels) should have tests that check : # rows < 32, # rows == 32, # rows > 32 and # rows > 64.   Lots of ways to flub those in kernels.

- For hand-rolled kernels, large enough inputs that they span multiple grid blocks.

- For anything involving strings or lists,  exhaustive combinations of empty strings/lists, null strings/lists and strings/lists with null elements.  

- Tests that explicitly use sliced columns.  That is, columns that have an offset > 0.   An easy to forget case.

- Tests that verify against various forms of "degenerate" column inputs.  Some typical cases are:  empty string columns that have no children (not many paths in cudf can generate these but it does happen);  columns that are of size 0 but somehow have non-null data pointers;  struct columns with no children.

- Decimal types are _not_ included with NumericTypes (but they are included with FixedWidthTypes). Be careful here.

Probably a decent number of other examples.

harrism replied:
Done. Thanks!

nvdbaranec commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,977 @@
+of any type. Note also that not only is each list nullable (using the null mask of the parent), but
+each list element may be nullable. So you may have a lists column with null row 3, and also null
+element 2 of row 4.
+
```
A key point with list columns is:
- The actual underlying data for a lists column is always bundled together into a single leaf column at the very bottom of the hierarchy (ignoring structs, which conceptually "reset" the root of the hierarchy), regardless of the level of nesting.  So a `List<List<List<List<int>>>>>` column still just has a single int column at the very bottom.  

A semi-visual representation might be useful here:

```
   lists_column = { {{{1, 2}, {3, 4}}, NULL}, {{{10, 20}, {30, 40}}, {{50, 60, 70}, {0}}} }

   List<List<List<int>>>  (2 rows):
   Length : 2
   Offsets : 0, 2, 4
   Children :
      List<List<int>>:
      Length : 4
      Offsets : 0, 2, 2, 4, 6
      Null count: 1
        1101
      Children :
        List<int>:
        Length : 6
        Offsets : 0, 2, 4, 6, 8, 11, 12
        Children :
          Column of ints
          1, 2, 3, 4, 10, 20, 30, 40, 50, 60, 70, 0
```

nvdbaranec replied:
A link to the arrow docs would be good too:

https://arrow.apache.org/docs/format/Columnar.html?highlight=nested%20types#physical-memory-layout

Our list implementation is specifically the "Variable-size List" physical layout.

harrism replied:
Added all of this. Thanks.

nvdbaranec commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,977 @@
+
+### View
+
+A view is a non-owning object that provides zero-copy access (possibly with slicing or offsets) data 
```
I think a section on slicing and column_view offsets is warranted.  It's an easy-to-overlook case and the word "offset" is heavily overloaded in cudf.

harrism replied:
Added

vuule commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,977 @@
+# cuIO: file reading and writing
+
+cuIO is a component of libcudf that provides GPU-accelerated reading and writing of data file 
+formats commonly used in data analytics, including CSV, Parquet, ORC, and Avro.
```
and JSON Lines (not sure if omitted intentionally).

harrism replied:
Added

vuule commented on cpp/docs/BENCHMARKING.md:
```diff
@@ -0,0 +1,44 @@
+# Unit Benchmarking in libcudf
```
Do we want to mention the benchmark utilities (currently only includes the random data generator)?

harrism replied:
I discussed the CUDA timer utility. I'm not aware of a random data generator. I can't find it in the benchmarks directory, can you point me to it? If you wrote the utility, perhaps you can write it up? Or maybe as a followup PR?

vuule replied:
I even presented it at a "better code" meeting :(
I can write something up in a followup PR, since this one is almost in.

harrism replied:
I forgot. In any case, can you point me to the files?

codereport commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,977 @@
+## Namespaces
+
+### External
+All public libcudf APIs should be placed in the `cudf` namespace*. Example:
```
```suggestion
All public libcudf APIs should be placed in the `cudf` namespace. Example:
```

codereport commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,977 @@
+than explicit `column`/`table`/`scalar` parameters. As with STL, iterators enable generic 
+algorithms to be applied to arbitrary containers. A good example of this is `cudf::copy_if_else`. 
+This function takes two inputs, and a Boolean mask. It copies the corresponding element from the 
+first or second input depending on whether the mask at that index is true or false. Implementing
```
```suggestion
first or second input depending on whether the mask at that index is `true` or `false`. Implementing
```

codereport commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,977 @@
+
+### Null-replacement iterator
+
+This iterator replaces the null/validity value for each element with a specified constant (true or
+false). Created using `cudf::detail::make_null_replacement_iterator`.
+
+### Validity iterator
+
+This iterator returns the validity of the underlying element (true or false). Created using 
```
```suggestion
This iterator replaces the null/validity value for each element with a specified constant (`true` or
`false`). Created using `cudf::detail::make_null_replacement_iterator`.

### Validity iterator

This iterator returns the validity of the underlying element (`true` or `false`). Created using 
```

codereport reviewed (changes requested):
Looking great! Just a few small things.

mythrocks commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,977 @@
+
+### Table
+
+A table is a collection of columns. It is also known as a dataframe.
```
> the C++ corollary to a DataFrame...

That is very interesting phrasing. I would've have prosaically termed it as "equivalent" to a `DataFrame`.

codereport replied:
i was just copying wording from later in the doc

harrism replied:
Hmmm, "corollary" doesn't appear later in the doc...

harrism replied:
I also wouldn't CamelCase or capitalize "data frame".  I decided to go with "equivalent", and fix spelling to "data frame".

mythrocks commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,977 @@
+creating a stream can be expensive. RMM has recently added a stream pool for this situation to 
+avoid dynamic stream creation. However, this is not yet exposed in libcudf, so for the time being, 
+libcudf features should avoid creating streams (even if it is slightly less efficient). It is a 
+good idea to leave a `// TODO:` note indicating where using a stream would be beneficial.
```
That `TODO` is intentional and valid, methinks.

mythrocks commented on cpp/docs/DEVELOPER_GUIDE.md:
````diff
@@ -0,0 +1,977 @@
+}
+```
+
+**Note:** It's important to synchronize the stream if *and only if* it is necessary. For example,
````
```suggestion
**Note:** It is important to synchronize the stream if *and only if* it is necessary. For example,
```

mythrocks commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,977 @@
+
+There may be times in implementing libcudf features where it would be advantageous to use streams 
+*internally*, i.e., to accomplish overlap in implementing an algorithm. However, dynamically 
+creating a stream can be expensive. RMM has recently added a stream pool for this situation to 
```
```suggestion
creating a stream can be expensive. RMM has a stream pool for this situation to 
```

mythrocks replied:
Hmm... Please feel free to ignore this suggestion. I wanted to avoid references to points in time, to prevent said reference from going out of date. But I see that the following sentence refers to this as `not yet exposed in libcudf`.

harrism replied:
I generally agree with your sentiment here. I'll improve it.

mythrocks commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,977 @@
+
+The pair iterator is used to access elements of nullable columns as a pair containing an element's 
+value and validity. `cudf::detail::make_pair_iterator` can be used to create a pair iterator from a 
+`column_device_view` or a `cudf::scalar`.
```
Should we mention explicitly that `make_pair_iterator` is not available on `mutable_column_device_view`?

mythrocks commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,977 @@
+libcudf supports a number of nested data types, including strings, lists, and structs. A string is 
+simply a character string, but a column of strings may have a different-length string in each row.
+A list is a list of elements of any type, so a column of lists of integers has rows with a list of 
+integers, possibly of a different length, in each row. In a colum of structs, each row is a 
```
```suggestion
integers, possibly of a different length, in each row. In a column of structs, each row is a 
```

mythrocks commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,977 @@
+
+# Nested Data Types
+
+libcudf supports a number of nested data types, including strings, lists, and structs. A string is 
+simply a character string, but a column of strings may have a different-length string in each row.
+A list is a list of elements of any type, so a column of lists of integers has rows with a list of 
+integers, possibly of a different length, in each row. In a colum of structs, each row is a 
+structure comprising one or more fields. These fields are stored in structure-of-arrays format, so 
+that the column of structs has a nested column for each field of the structure. 
```
Might we consider converting this paragraph to a list? (No pun intended.) 
Also, fixed typo in `colum`.
```suggestion
libcudf supports a number of nested data types, including strings, lists, and structs: 
1. `String`: This is simply a character string, but a column of strings may have a 
different-length string in each row.
2. `List`: This is a list of elements of any type, so a column of lists of integers has 
rows with a list of integers, possibly of a different length, in each row. 
3. `Struct`: In a column of structs, each row is a structure comprising one or more 
fields. These fields are stored in structure-of-arrays format, so that the column of 
structs has a nested column for each field of the structure. 
```

mythrocks commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,977 @@
+
+# Nested Data Types
+
+libcudf supports a number of nested data types, including strings, lists, and structs. A string is 
```
I wonder if we are conflating "variable size" types with "nested types" here:
1. Are strings really nested types?
2. `struct` is only of variable size if it (eventually) contains a string or list.

nvdbaranec replied:
>     1. Are strings really nested types?

This is a really slippery question.  To me, they are in limbo between nested and non-nested.  

The string column itself does indeed have child columns "nested" inside but that's more of an implementation detail than a data type. It's still logically just "strings".   On the other hand, 1 level list columns are exactly the same and those are clearly nested types.  

On the other, _other_ hand, I've found that when writing tests, I tend to only use the term "Nested" in the name when the test in question has >= two logical columns nested inside each other (ie, List<List<int>>).  If I'm just making a test with a basic list or struct, I tend to use "List" or "Struct" in the name.  

Arrow probably has something to say about this.

harrism replied:
I rephrased to "Variable-size and Nested Data Types".

mythrocks commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,977 @@
+
+Structs are represented similarly to lists, except that they have multiple child data columns.
+The parent column's type is `STRUCT` and contains no data, but its size represents the number of 
+structs in the column, and its null mask represents the validity of each struct element. The parent 
```
I'm trying to decide whether the interplay between the validity masks of structs and its members should be described here. Specifically that the parent's validity mask is bitwise-ANDed with all its children's masks, so that a null `struct` row at the parent level marks the corresponding child row to null, for all child columns.

Would an explanation (with examples)  make this section too verbose?

harrism replied:
It might, but it's better to be too verbose and solve that problem by refactoring than to have missing information. If you can provide something in a suggestion comment, I'll accept it.

harrism replied:
I'm not sure I would get this written correctly by myself. Would be great if you could write it, either in this PR or file an issue after and do it as a followup.

mythrocks replied:
Sorry I missed this. I'll add a section shortly.

mythrocks replied:
Hmm... I can't seem to add a `suggestion` to this diff. 
Perhaps something like this? (Will likely need rephrasing.)
Please note that the "\\```" will need the "\\" removed. I couldn't figure out how to escape this correctly in the suggestion text. :/

```
As per the [Apache Arrow specification](https://arrow.apache.org/docs/format/Columnar.html#struct-layout), in addition to the struct column's null mask, each struct
field column has its own optional null mask. A struct field's validity can vary independently from the 
corresponding struct row. For instance, a non-null struct row might have a null field. However, the fields
of a null struct row are deemed to be null as well.
As an example, consider a struct column of type `STRUCT<FLOAT32, INT32>`. If the contents were
`[ {1.0, 2}, {4.0, 5}, null, {8.0, null} ]`, the struct column's layout would be as follows:

\```
{
  type = STRUCT
  null_mask = 1011
  null_count = 1
  children = {
    {   
      type = FLOAT32
      data = [1.0, 4.0, X, 8.0]
      null_mask  = 1011
      null_count = 1
    },  
    {   
      type = INT32
      data = [2, 5, X, X]
      null_mask  = 0011
      null_count = 2
    }  
  }   
}
\```
The last struct row (index 3) is not null, but has a null value for its INT32 field. This is 
reflected in bit 3 of the null masks of the struct column and the INT32 field.
Also, row 2 of the struct column is null. Its corresponding fields should also be null, so bit 2 is
unset in the null masks of both struct fields.

```

davidwendt replied:
I believe the bit indices are shown right-to-left `76543210`. 
So `1011` means that only `row0`, `row1`, and `row3` are valid.

mythrocks replied:
1 == valid. 0 == null.
I based the null-mask representation on how it shows up when the column is printed. I realize that this is confusing.

I have modified the example to show the null_mask as an array instead. This should be clearer.

mythrocks replied:
@davidwendt, et al, 

Would anyone know how one makes a "```suggestion" block that itself contains a "```" code block, in GitHub? I'd like to make my suggestion above available to pick into this PR, but I'm thwarted by GitHub's formatting. :/

davidwendt replied:
It looks like you may be able to change the 3 backtick marks to 3 tildes for the suggestion block.
`````
~~~suggestion
Some text.
```
Example text.
```
Other text.
~~~
`````

I think it is also possible to nest blocks by using more than 3 backticks.

harrism replied:
@mythrocks the problem I have with your example is this: 

> The last struct row (index 3) is not null, but has a null value in the INT32 field. This is reflected in bit 3 of the null masks of the struct column and the INT32 field.

But the column-level (parent-level?) null mask is [1, 1, 0, 1] (bit 3 is valid), while the INT32 field null mask is [1, 1, 0, 0] (bit 3 is invalid/null).  So your text does not match your example.

mythrocks replied:
>> The last struct row (index 3) is not null...

> But the column-level (parent-level?) null mask is [1, 1, 0, 1] (bit 3 is valid)

A) `1` indicates valid. In the `STRUCT` mask, bit 3 is 1. Therefore, `STRUCT` at index 3 is valid, i.e. not null. The two quotes above are not at odds.

>> but has a null value in the INT32 field...

> while the INT32 field null mask is [1, 1, 0, 0] (bit 3 is invalid/null).

B) So the `INT32` field's mask[3] is 0, i.e. invalid.

From A and B above, the `STRUCT` is valid, but its `INT32` field is invalid. Hence, `{8.0, ∅}`.

Forgive me, I'm still unable to see the problem. :/

harrism replied:
Sorry, I read 

> This is reflected in bit 3 of the null masks of the struct column and the INT32 field.

To mean that the null masks of the struct column *and* the INT32 field should be null at bit 3. I will rewrite this sentence to be clear, it's unclear currently.

mythrocks replied:
Apologies for the clumsy wording. I meant that "this" is reflected in bit 3 of the null masks of the `STRUCT` column and the `INT32` field (i.e. `1` and `0` respectively).

davidwendt commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,977 @@
+machine to use the `cudf/cpp/.clang-format` configuration file, and run `clang-format` on all 
+changed code before committing it. The easiest way to do this is to configure your editor to 
+"format on save".
+
```
How about some guidelines for `#include` statements? 
From @harrism slack message in November:

1. Group your includes by library (e.g. cuDF, RMM, Thrust, STL)
2. Separate groups by a blank line.
3. Order the groups from "nearest" to "farthest"
4. Use `<>` instead of `""` unless the header is in the same directory as the source file.
5. Let clang-format sort the headers within the groups
6. If you remove code, check whether that makes some includes obsolete, and remove them if so.

Maybe 6 can be generalized with something like:
"Always check the included files are only necessary for the file in which they are included. Try to avoid excessive including especially in header files."

harrism replied:
Added.

davidwendt commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,977 @@
+
+Aspects of code style not discussed above and not automatically enforceable are typically caught
+during code review, or not enforced.
+
```
Anything else we should include here?

- Usage of `static_cast` and `reinterpret_cast` instead of cast operator `(type)`
- Avoiding the comma operator https://github.com/rapidsai/cudf/pull/7090#discussion_r553330672
- Prefer using thrust or STL algorithms over for-loops when possible.

harrism replied:
I could just link the C++ core guidelines and Sean Parent's C++ seasoning talk and say "commit this to memory". :)

harrism replied:
Added a C++ Guidelines subsection. Thanks!

davidwendt commented on cpp/docs/DEVELOPER_GUIDE.md:
````diff
@@ -0,0 +1,977 @@
+// This will always invoke `operator()<int32_t>`
+cudf::type_dispatcher<always_int>(data_type, f);
+```
+
````
I feel we need a Spider-Man quote here: _With great power comes great responsibility._

Try to avoid multiple type-dispatch if possible. The compiler will create a code path for every type so a 2nd type-dispatch would multiply the paths by the number of types. This can create very large object code and cause very long compile times.

harrism replied:
Added

davidwendt commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,1060 @@
+and we try to follow his rules: "No raw loops. No raw pointers. No raw synchronization primitives." 
+
+ * Prefer algorithms from STL and Thrust to raw loops.
+ * Prefer libcudf and RMM [owning data structures and views](libcudf-data-tructures) to raw pointers
```
```suggestion
 * Prefer libcudf and RMM [owning data structures and views](libcudf-data-structures) to raw pointers
```

codereport commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,1060 @@
+|-|-|-|
+|fixed-width|`fixed_width_scalar<T>`| `T` can be any fixed-width type|
+|numeric|`numeric_scalar<T>` | `T` can be `int8_t`, `int16_t`, `int32_t`, `int_64_t`, `float` or `double`|
+|fixed-point|`fixed_point_scalar<T> | `T` can be `numeric::decimal32` or `numeric::decimal64`|
```
```suggestion
|fixed-point|`fixed_point_scalar<T>` | `T` can be `numeric::decimal32` or `numeric::decimal64`|
```

harrism replied:
Great catch.

codereport reviewed (changes requested):
Looks awesome! One small ` missing

nvdbaranec reviewed (approved):
This doc is excellent.

vuule commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,1060 @@
+In general, we recommend following 
+[C++ Core Guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines). We also 
+recommend watching Sean Parent's [C++ Seasoning talk](https://www.youtube.com/watch?v=W2tWOdzgXHA), 
+and we try to follow his rules: "No raw loops. No raw pointers. No raw synchronization primitives." 
```
AFAIK the pointer rule only refers to owning pointers. Passing pointers around without changing the ownership should be fine.
```suggestion
and we try to follow his rules: "No raw loops. No raw owning pointers. No raw synchronization primitives." 
```

harrism replied:
But it's a direct quote. And I think my bullet point after the quote clears this up. In general I don't think I agree. You will notice that we don't have raw pointers at all in our APIs unless necessary - this is what our view classes help with.

vuule replied:
Did not realize it's literally a quote, disregard the comment then. Just trying to make the guideline clearer. 
Further down in the document we do recommend to pass in/out parameters as pointers (unless it's a type with a view).

vuule commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,1060 @@
+### Multiple Return Values
+
+Sometimes it is necessary for functions to have multiple outputs. There are a few ways this can be 
+done in C++ (including creating a  `struct`  for the output). One convenient way to do this is 
```
Not sure if intentional
```suggestion
done in C++ (including creating a `struct` for the output). One convenient way to do this is 
```

vuule commented on cpp/docs/DEVELOPER_GUIDE.md:
````diff
@@ -0,0 +1,1060 @@
+```
+
+Note:  `std::tuple`  _could_  be used if not for the fact that Cython does not support 
+`std::tuple`. Therefore, libcudf APIs must use  `std::pair`, and are therefore limited to return 
````
```suggestion
`std::tuple`. Therefore, libcudf APIs must use `std::pair`, and are therefore limited to return 
```

vuule reviewed (commented):
Have to admit, learned a lot from reviewing this :)
Found a couple of small typos.

codereport reviewed (approved):
:fire:

davidwendt commented on cpp/docs/TESTING.md:
```diff
@@ -0,0 +1,447 @@
+   but also applies to Thrust and CUB algorithm calls with lambdas / functors. 
+
+ * For anything involving strings or lists, test exhaustive combinations of empty strings/lists,
+   null strings/lists and strings/lists with null elements.
```
This came up today in a review.

"Also, for strings tests, include a mixture of non-ASCII UTF-8 characters like é in your test strings."

harrism commented:
@gpucibot merge
